Realm/Realm/Transaction.cs

```diff
@@ -55,30 +70,80 @@ public void Dispose()
         /// </summary>
         public void Rollback()
         {
-            if (_realm == null)
+            EnsureActionFeasibility();
+
+            if (!IsAsync())
             {
-                throw new Exception("Transaction was already closed. Cannot roll back");
+                _realm.SharedRealmHandle.CancelTransaction();
+            }
+            else
+            {
+                _realm.SharedRealmHandle.CancelAsyncTransaction(_asyncTransactionHandle.Value);
             }
 
-            _realm.SharedRealmHandle.CancelTransaction();
-            _realm.DrainTransactionQueue();
-            _realm = null;
+            FinishTransaction();
         }
 
         /// <summary>
         /// Use to save the changes to the realm. If <see cref="Transaction"/> is declared in a <c>using</c> block,
         /// must be used before the end of that block.
         /// </summary>
         public void Commit()
+        {
+            EnsureActionFeasibility();
+            _realm.SharedRealmHandle.CommitTransaction();
+            FinishTransaction();
+        }
+
+        public async Task CommitAsync()
+        {
+            EnsureActionFeasibility();
+            _asyncTransactionHandle = await _realm.SharedRealmHandle.CommitTransactionAsync();
+            FinishTransaction();
+        }
+
+
+        // TODO andrea: check usefulness of this method
+        private bool IsAsync()
```

Comment:
Do you see any value in the `Debug.Assert` in this method? I thought it'd be a good idea to have it around. If you do see value, where would you recommend to have it? At the beginning of the `Transaction.Rollback`, perhaps?

Reply:
No, I don't see much value in debug asserting here - this will get optimized away when we compile, so it won't reach user code. I expect Core already validates this, but if it doesn't, we can throw an exception in the wrappers for commit/rollback

Reply:
I only meant that `Debug.Assert` for our development only. Mostly to ensure correctness in the way managed keeps track of `_asyncTransactionHandle`.
I was more worried about our side than core :D. I will just take that part out.
Thanks.